File: src/java_version/com/common/SymbolTable.java
package com.common;

import java.util.HashMap;

public class SymbolTable
{
    private SymbolTable  prev = null;
    private HashMap<String, Symbol> table;
    private int level = 0;

    public SymbolTable(SymbolTable prev)
    {
        this.prev = prev;
        if (this.prev != null)
        {
            this.level = prev.level + 1;
        }
        this.table = new HashMap<String, Symbol>();
    }

    public void add(Symbol symbol)
    {
        if (this.table.containsKey(symbol.name))
        {
            Parser.error(19);
        }else
        {
            this.table.put(symbol.name, symbol);
        }
    }

    public Symbol get(String name)
    {
        SymbolTable curr = this;
        while (curr != null)
        {
            if (curr.table.containsKey(name))
            {
                return curr.table.get(name);
            }
            curr = curr.prev;
        }
        return null;
    }

    public int getLevel()
    {
        return this.level;
    }
}

File: src/SymbolTable.cpp
//
// Created by manout on 17-11-5.
//

#include "SymbolTable.h"
SymbolTable::SymbolTable(SymbolTable *prev):table(), prev(prev)
{
	if (prev not_eq nullptr)
	{
		level = prev->level + 1;
	}
	else
	{
		level = 0;
	}
}
SymbolTable &SymbolTable::add(const Symbol & item)
{
	if (this->table.find(item.name) == this->table.end())
	{
		this->table[item.name] = item;
	}
	else
	{
		error(19);
	}
	return *this;
}

SymbolTable &SymbolTable::add(const Symbol* item_ptr)
{
	this->add(*item_ptr);
	return *this;
}

Symbol *SymbolTable::get(const std::string& id)
{
	for (SymbolTable* curr_table = this;
	        curr_table not_eq nullptr;
				curr_table = curr_table->prev)
	{
		auto query_ret = curr_table->table.find(id);
		if (query_ret not_eq curr_table->table.end())
		{
			return &query_ret->second;
		}
	}
	return nullptr;
}

int SymbolTable::get_level()
{
	return this->level;
}


